libraries/standard/http/src/http_client.c

```diff
@@ -247,40 +307,203 @@ static HTTPStatus_t _sendHttpBody( const HTTPTransportInterface_t * pTransport,
+
+    /* A transport status of less than zero is an error. */
+    if( transportStatus < 0 )
+    {
+        IotLogErrorWithArgs( "Failed to receive HTTP response: Transport recv() "
+                             "returned error: TransportStatus=%d.",
+                             transportStatus );
+        returnStatus = HTTP_NETWORK_ERROR;
+    }
+    else if( (size_t)transportStatus > bufferLen )
+    {
+        /* There is a bug in the transport recv if more bytes are reported
+         * to have been read than the bytes asked for. */
+        IotLogErrorWithArgs( "Failed to receive HTTP response: Transport recv() "
+                             " read more bytes than expected: BytesRead=%d",
+                             transportStatus );
+        returnStatus = HTTP_NETWORK_ERROR;
+    }
+    else if( transportStatus > 0 )
+    {
+        /* Some or all of the specified data was received. */
+        *pBytesReceived = ( size_t ) ( transportStatus );
+        IotLogDebugWithArgs( "Received data from the transport: BytesReceived=%d.",
+                             transportStatus );
+    }
+    else
     {
-        transportStatus = pTransport->send( pTransport->pContext,
-                                            pRequestBodyBuf,
-                                            reqBodyBufLen );
+        /* When a zero is returned from the transport recv it will not be
+         * invoked again. */
+        IotLogDebug( "Transport recv() returned 0. Receiving transport data"
+                     "is complete." );
+    }
+
+    return returnStatus;
+}
 
-        if( transportStatus < 0 )
+/*-----------------------------------------------------------*/
+
+static HTTPStatus_t _getFinalResponseStatus( HTTPParsingState_t parsingState,
+                                             size_t totalReceived,
+                                             size_t responseBufferLen )
+{
+    HTTPStatus_t returnStatus = HTTP_SUCCESS;
+
+    assert( parsingState >= HTTP_PARSING_NONE &&
+            parsingState <= HTTP_PARSING_COMPLETE );
+    assert( totalReceived <= responseBufferLen );
+
+    /* If no parsing occurred, that means network data was never received. */
+    if( parsingState == HTTP_PARSING_NONE )
+    {
+        IotLogErrorWithArgs( "Response not received: Zero returned from "
+                             "transport recv: totalReceived=%d",
+                             totalReceived );
+        returnStatus = HTTP_NO_RESPONSE;
+    }
+    else if( parsingState == HTTP_PARSING_INCOMPLETE )
+    {
+        if( totalReceived == responseBufferLen )
         {
-            IotLogErrorWithArgs( "Error in sending the HTTP body over the "
-                                 "transport interface. Transport status %d.",
-                                 transportStatus );
-            returnStatus = HTTP_NETWORK_ERROR;
+            IotLogErrorWithArgs( "Response is too large for the response buffer"
```

Comment:
"Unable to read complete response from transport interface: Response buffer has insufficent space: ResponseBufferSize=%d"

Reply:
sure